Update laser_ni.py

AdvancedImagingUTSW opened the pull request:
For the PCIE6738 card, we have an abundance of analog outputs, but not many digital outputs. So, we need to sometimes make an analog output act like a digital output. I have added some logic to make this possible. Works on the SpectralTIRF system.

AdvancedImagingUTSW committed:
Update laser_ni.py

For the PCIE6738 card, we have an abundance of analog outputs, but not many digital outputs. So, we need to sometimes make an analog output act like a digital output. I have added some logic to make this possible. Works on the SpectralTIRF system.

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -48,16 +48,26 @@
+
+        except KeyError as e:
             self.laser_do_task = None
             if isinstance(e, DaqError):
```
This check will always be `False` since we are excepting a `KeyError`.

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -48,16 +48,26 @@
+        self.on_off_type = None
 
         # Digital out (if using mixed modulation mode)
         try:
```
Try/excepting ao vs do feels dangerous. Could we create two classes of LaserNI instead, where we can explicitly specify which is which from the YAML file?

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -48,16 +48,26 @@
-                laser_do_port, line_grouping=LineGrouping.CHAN_FOR_ALL_LINES
-            )
-        except (KeyError, DaqError) as e:
+            try:
```
Try/excepting ao vs do feels dangerous. Could we create two classes of LaserNI instead, where we can explicitly specify which is which from the YAML file?

AdvancedImagingUTSW committed:
Update laser_ni.py

Now we perform a string search to see if it includes "/ao" in the port name. I feel like this should be fairly bombproof, and avoids any need to have yet another item in our configuration file.

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -48,15 +48,26 @@
         try:
             laser_do_port = self.device_config["onoff"]["hardware"]["channel"]
-
+            self.laser_min_do = self.device_config["onoff"]["hardware"]["min"]
```
Since we do not need this information for digital channels, I think we should move this into the `if "/ao"` control flow or make this a get or default.

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -48,15 +48,26 @@
 class LaserNI(LaserBase):
     def __init__(self, microscope_name, device_connection, configuration, laser_id):
         super().__init__(microscope_name, device_connection, configuration, laser_id)
+        self.on_off_type = None
```
What about changing this to `self.on_off_type = "analog" if "/ao" in laser_do_port else "digital"` and then using this flag to make the decisions? Better yet, `self.on_off_analog = True if "/ao" in laser_do_port else False` and then the rest of the checks are just against a `bool` instead of a full string comparison (faster).

zacsimile commented on src/aslm/model/devices/lasers/laser_ni.py:
```diff
@@ -98,7 +108,10 @@ def set_power(self, laser_intensity):
+                if self.on_off_type == "digital":
+                    self.laser_do_task.write(True, auto_start=True)
+                elif self.on_off_type == "analog":
+                    self.laser_do_task.write(self.laser_max_do, auto_start=True)
```
If you make `self.laser_max_do = self.device_config["onoff"]["hardware"].get("max", 1)` and `self.laser_min_do = self.device_config["onoff"]["hardware"].get("min", 0)` you can use Python's duck typing to automatically write `True` (`1`)/`False` (`0`) in the absence of specified min and max values and skip the `if self.on_off_type` checks entirely.
